### PR TITLE
 drivers: spi: esp32: add option to handle lines state

### PR DESCRIPTION
**Pull request by sylvioalves**
SPI driver is current working for common SPI devices. However, addressable LED like WS2812 requires MOSI line to be default LOW during initialization. This PR adds such option in SPI configuration. This has no effect on common SPI operation.

### CODE REVIEW

**Comment by sylvioalves** [comment]
@marekmatej I think this will conflict with SoC refactoring PR, right?

**Review by henrikbrixandersen — COMMENTED** [review]
This may have an impact on power consumption for some boards. It's quite common to add external pull-ups to SPI lines. Could this be handled through pinctrl in any way?

**Comment by EricNRS** [comment]
> This may have an impact on power consumption for some boards. It's quite common to add external pull-ups to SPI lines. 
Thank you for thinking about power consumption.  SPI normally uses push-pull drive, so pull-ups aren't used, although I have seen some engineers put a pull-up or pull-down on the MISO line to prevent the line from floating when a slave isn't selected.  Any chance you are thinking about I2C where pull-ups are required?

For SPI, inactive low (which is what this change does) is normally what I am used to as a default.  Still, I share your concern that this could break some existing applications if we do not make it configurable.  Would an extra bit in the `struct spi_config::operation` field (include/zephyr/drivers/spi.h) be appropriate for this configuration since there is already the clock polarity bit (SPI_MODE_CPOL)?

Just out of curiosity, is there any reason why the bit definitions such as `SPI_MODE_CPOL` are in `zephyr/include/zephyr/drivers/spi.h` and not in `zephyr/include/zephyr/dt-bindings/spi/spi.h`?  Seems like they should be moved to the DT-bindings file and then included in `zephyr/include/zephyr/drivers/spi.h` to allow the operation field to be configured in the DT.

**Comment by marekmatej** [comment]
> @marekmatej I think this will conflict with SoC refactoring PR, right?

Yes, but the conflict is minor, no big deal.

**Comment by sylvioalves** [comment]
> This may have an impact on power consumption for some boards. It's quite common to add external pull-ups to SPI lines. Could this be handled through pinctrl in any way?

@henrikbrixandersen, I don't see a way to perform that in pinctrl level. Even if I bring the MOSI GPIO to low, once SPI is gets configured, it automatically brings MOSI line high for all ESP32-XX by default. ESP32 keeps MOSI line LOW by default though.  So, my considerations:  
1) For some reason, WS2812 LED Strip does not work properly when MOSI line starts in high level. TBH I have not checked WS2812/SPI implementation. Should it force to low prior any operation somehow?  
2) For common SPI operations, MOSI line idle state is pointless as CPOL/CPHA defines its operation.  
3) For this change, I see only the option to add some ESP32 custom SPI config in DTS or updating base SPI struct with a new bit. Any recommendation?

Thanks!

**Comment by EricNRS** [comment]
> 1. For some reason, WS2812 LED Strip does not work properly when MOSI line starts in high level. TBH I have not checked WS2812/SPI implementation. Should it force to low prior any operation somehow?

The WS2812 requires MOSI to be low for >=80us after receiving data to latch the data and reset the shift register for a new command.  Each PMW bit (which is an 8-bit SPI byte) takes 1.2us to transfer so it would require 80/1.2 = 66.6 ==> 67 bytes of 0x00 to be transmitted before and after the actual LED bit data to ensure correct behavior.  This is an excessive amount of overhead IMHO, so best to add a configuration for the data polarity bit.  Since MOSI being low when not active is what I am used to on other hardware, I would suggest that this is the default for the ESP32 targets as well and that the polarity can be changed using the API if anyone requires it.

**Comment by EricNRS** [comment]
> I looked on the `WS2812*` portfolio, and the reset time per model varies significantly. For example the `WS2812B-*` module has reset time >280us, while `WS2812` states >50us. Therefore I agree with @EricNRS that a polarity bit is probably the best solution.
> 
> * Note that all the `esp32s[23]_devkit[cm]` boards are populated by the `WS2812B-Mini` modules.

Thank you for inspecting the datasheets -- quite the variation in reset timing!

**Comment by sylvioalves** [comment]
Folks, PTAL.

**Comment by EricNRS** [comment]
> Folks, PTAL. This approach fits as we are targetting only ESP32 devices?

Works for me.  If we see this on other chipsets in the future, then we can look at some of the other suggested options to make this a general setting.

**Comment by sylvioalves** [comment]
I'll add the led_ws1812 sample code for ESP32-S3 devkitm adding this config into spi overlay.

**Comment by simonguinot** [comment]
Since https://github.com/zephyrproject-rtos/zephyr/pull/46936/commits/2598329c8eb3411207b0d7e501c5fb768fa8608a the `spi-cpol` and `spi-cpha` options can be used to control signal between transmissions. Maybe this can help.

**Comment by marekmatej** [comment]
Folks, so I've been playing with the WS2812 led, which is populated on most of ESP32 boards I have (except for basic ESP32).
I tried changing the ws2912_spi.c to test the approach with an extended LOW period before each pixel transfer. Even the 40 bytes of '0' was not sufficient to make the chip understand the input sequence afterward. Increasing the 'zero-filled-buffer' even more is causing significant overhead and it is not acceptable for the string of the WE2812.
Driving the MOSI low is the only solution that worked for me on the ESP32S3, ESP32S3 and ESP32C3 based boards.

It is also worth mentioning that minor overlay changes (MOSI assignment on `pinctrl`) were needed to make the ws2812 sample code work. Also note, that `esp32s3_devkitc` v1.0 has RGB led connected to GPIO48, while v1.1 has it on GPIO38.

File: dts/bindings/spi/espressif,esp32-spi.yaml [hunk 1]
```diff
@@ -76,3 +76,8 @@ properties:
     description: |
       Chip select hold time setting, see TRF for SOC for details of
       timing applied.
+
+  line-idle-low:
+    type: bool
```
Comment (beriberikix) [inline dts/bindings/spi/espressif,esp32-spi.yaml:81]: Should this be `boolean`? I ran into a compile error.

```
devicetree error: 'line-idle-low' in 'properties:' in C:/Users/jberi/code/zephyrproject/zephyr/dts/bindings\spi\espressif,esp32-spi.yaml has unknown type 'bool', expected one of boolean, string-array, int, phandle-array, compound, phandles, phandle, uint8-array, path, array, string
```

File: drivers/spi/spi_esp32_spim.c [hunk 2]
```diff
@@ -347,6 +347,17 @@ static int IRAM_ATTR spi_esp32_configure(const struct device *dev,
...
+	/* Workaround to make MOSI and MISO line low as default as there is not API for that */
+#ifndef CONFIG_SOC_ESP32
+	if (hal_dev->line_idle_low) {
+		spi_dev_t *hw = hal->hw;
+		hw->ctrl.d_pol = 0;
+		hw->ctrl.q_pol = 0;
+	}
+#endif
+
+#endif
```
Comment (beriberikix) [inline drivers/spi/spi_esp32_spim.c:359]: Extra `endif`?

**Comment by beriberikix** [comment]
Hi @sylvioalves! Thanks for working on this :) I'm trying to test it out and ran into some issues and left them as comments. The last one I'm working through is around the hal config:

```
C:/Users/jberi/code/zephyrproject/zephyr/drivers/spi/spi_esp32_spim.c: In function 'spi_esp32_configure':
C:/Users/jberi/code/zephyrproject/zephyr/drivers/spi/spi_esp32_spim.c:352:20: error: 'spi_hal_dev_config_t' has no member named 'line_idle_low'
  352 |         if (hal_dev->line_idle_low) {
      |                    ^~
```

**Comment by sylvioalves** [comment]
@beriberikix Hey Jonathan, sorry, this PR was supposed to be a draft. I was on leave and just submitted this to evaluate opinions about the approach. Now that I am back all continue handling this. Anyway, thanks.

**Comment by beriberikix** [comment]
All good! Happy to test when ready!

**Review by beriberikix — APPROVED** [review]
I was able to successfully test this on both a `esp32c3_devkitm` (-1 v1.0) & `esp32c3_devkitc` (-02 v1.1.) I used the sample overlays from [60520](https://github.com/zephyrproject-rtos/zephyr/pull/60520) but had to add `line-idle-low;` to the `spi` entry.

**Review by henrikbrixandersen — COMMENTED** [review]
I am still not getting this. How does this co-exist with `SPI_MODE_CPOL` and `SPI_MODE_CPHA`? Shouldn't the SPI controller driver be set up by the SPI peripheral driver to use the correct idle polarity?

**Comment by EricNRS** [comment]
> I am still not getting this. How does this co-exist with `SPI_MODE_CPOL` and `SPI_MODE_CPHA`? Shouldn't the SPI controller driver be set up by the SPI peripheral driver to use the correct idle polarity?

CPOL and CPHA are for the clock polarity and phase.  This change controls what the data line does when there isn't a transfer in progress (e.g. idle bus, no slave selected).  I have never seen this configuration on other SPI peripherals, so it seems to be unique to Espressif.  A lot of times, MOSI is directly driven by the output shift register, so the idle state would just be the last bit sent over the bus.

All of this likely doesn't matter for real SPI slave devices, but it does matter for repurposing the SPI peripheral for sending PWM bits for addressable LEDs like the WS2812.

File: drivers/spi/spi_esp32_spim.c [hunk 3]
```diff
@@ -505,6 +517,7 @@ static const struct spi_driver_api spi_api = {
 		.dma_host = DT_INST_PROP(idx, dma_host),	\
 		.cs_setup = DT_INST_PROP_OR(idx, cs_setup_time, 0), \
 		.cs_hold = DT_INST_PROP_OR(idx, cs_hold_time, 0), \
+		.line_idle_low = DT_INST_PROP_OR(idx, line_idle_low, 0), \
```
Comment (henrikbrixandersen) [inline drivers/spi/spi_esp32_spim.c:520]: Boolean DTS properties will always be defined. 
```suggestion
		.line_idle_low = DT_INST_PROP(idx, line_idle_low), \
```

**Review by henrikbrixandersen — CHANGES_REQUESTED** [review]
Please update the PR description as this (no longer?) makes the lines low by default, but only if the given DTS property is set.

**Review by henrikbrixandersen — APPROVED** [review]

**Comment by sylvioalves** [comment]
@carlescufi @cfriedt PTAL.

**Review by cfriedt — APPROVED** [review]
LGTM

**Pull request merged**